csrc/iter_visitor.cpp

```diff
@@ -933,4 +933,162 @@ std::vector<Val*> InputsOf::outputs(
   return io.ordered_inputs;
 }
 
+/* DEAD CODE REMOVER */
+void DeadCodeRemover::run() {
+  // First we build a set of all live Statements so that we can detect dead
+  // branches.
+  for (auto stmt : StmtSort::getStmts(fusion_, fusion_->outputs())) {
+    markLive(stmt);
+  }
+
+  // Note that StmtSort::getStmts() is also run in traverseTo. In the future,
+  // we could potentially refactor this so that derived classes from
+  // BackwardVisitor can make use of that traversal instead of repeating it.
+  traverseTo(fusion_, fusion_->outputs());
+
+  // We do not remove Statements from the Fusion while traversing, to avoid
+  // dereferencing invalid pointers. Instead, we wait until this point to do the
+  // removal.
+  doRemoval();
+}
+
+Fusion* DeadCodeRemover::fusion() const {
+  return fusion_;
+}
+
+void DeadCodeRemover::handle(Statement* stmt) {
+  if (isDead(stmt)) {
+    // We check whether stmt is dead before we dereference it, since it may
+    // have been removed from the Fusion.
+    return;
+  }
+  BackwardVisitor::handle(stmt);
+}
+
+void DeadCodeRemover::handle(Expr* expr) {
+  if (maybeRemoveExpr(expr)) {
+    // maybeRemoveExp will remove expr from the Fusion if all its uses are
+    // marked dead. In that case, we should not continue handling it since the
+    // expr pointer is invalid.
+    return;
+  }
+  BackwardVisitor::handle(expr);
+}
+
+bool DeadCodeRemover::replaceVal(Val* old_val, Val* new_val) {
+  if (old_val->isFusionOutput()) {
+    fusion_->replaceOutput(old_val, new_val);
+  }
+  for (auto use : old_val->uses()) {
+    ir_utils::replaceValInExpr(use, old_val, new_val);
+  }
```

Comment:
Doesn't this mean the fusion is being changed during a traversal? Is this a safe change?

Reply:
You're correct, and that feels wrong. I just pushed another commit that simplifies things further, and avoids this mutation during traversal. Now, the child class only needs to use `replaceVal(Val* old_val, Val* new_val)` (`removeVal` is no longer exposed). All actual replacing and removing is handled inside the parent class `DeadCodeRemover` just like before, but now the `replaceValInExpr` is called at the end, just before removing registered Vals. This way, it is safe to dereference any pointer during the traversal.